diff -r 438bd63a1ab8 lib/dot_draw/src/olsrd_dot_draw.c
--- a/lib/dot_draw/src/olsrd_dot_draw.c	Fri Oct 16 17:18:44 2009 +0200
+++ b/lib/dot_draw/src/olsrd_dot_draw.c	Mon Oct 19 13:19:07 2009 +0200
@@ -267,7 +267,7 @@
     return;
   }
 #ifndef _WRS_KERNEL
-  if (!ip4equal(&pin.sin_addr, &ipc_accept_ip.v4)) {
+  if (!ip4equal(&pin.sin_addr, &ipc_accept_ip.v4) && pin.sin_addr.s_addr != INADDR_ANY) {
     olsr_printf(0, "Front end-connection from foreign host (%s) not allowed!\n", inet_ntoa(pin.sin_addr));
     CLOSE(ipc_connection);
     return;
diff -r 438bd63a1ab8 lib/nameservice/src/nameservice.c
--- a/lib/nameservice/src/nameservice.c	Fri Oct 16 17:18:44 2009 +0200
+++ b/lib/nameservice/src/nameservice.c	Mon Oct 19 13:19:07 2009 +0200
@@ -35,6 +35,7 @@
 /*
  * Dynamic linked library for UniK OLSRd
  */
+#define _GNU_SOURCE 1
 
 #define _GNU_SOURCE 1
 
diff -r 438bd63a1ab8 lib/txtinfo/src/olsrd_txtinfo.c
--- a/lib/txtinfo/src/olsrd_txtinfo.c	Fri Oct 16 17:18:44 2009 +0200
+++ b/lib/txtinfo/src/olsrd_txtinfo.c	Mon Oct 19 13:19:07 2009 +0200
@@ -257,7 +257,7 @@
     sin4 = (struct sockaddr_in *)&pin;
     if (inet_ntop(olsr_cnf->ip_version, &sin4->sin_addr, addr, INET6_ADDRSTRLEN) == NULL)
       addr[0] = '\0';
-    if (!ip4equal(&sin4->sin_addr, &txtinfo_accept_ip.v4)) {
+    if (!ip4equal(&sin4->sin_addr, &txtinfo_accept_ip.v4) && sin4->sin_addr.s_addr != INADDR_ANY) {
       olsr_printf(1, "(TXTINFO) From host(%s) not allowed!\n", addr);
       close(ipc_connection);
       return;
